clients/src/main/java/org/apache/kafka/clients/consumer/internals/Acknowledgements.java

```diff
@@ -246,10 +287,181 @@ public List<ShareAcknowledgeRequestData.AcknowledgementBatch> getShareAcknowledg
                 }
             }
         }
-        batches.add(currentBatch);
+        List<ShareAcknowledgeRequestData.AcknowledgementBatch> optimalBatches = maybeOptimiseAcknowledgementTypesForAcknowledge(currentBatch);
+
+        optimalBatches.forEach(batch -> {
+            if (canOptimiseForSingleAcknowledgeType(batch)) {
+                // If the batch had a single acknowledgement type, we optimise the array independent
+                // of the number of records.
+                batch.acknowledgeTypes().subList(1, batch.acknowledgeTypes().size()).clear();
+            }
+            batches.add(batch);
+        });
+        return batches;
+    }
+
+    /**
+     * Traverses the shareFetchBatch and splits it into optimal batches wherever possible.
+     * Optimisation happens when a batch has continuous records of the same acknowledge type
+     * whose count exceeds the default value. In this case, the batch is split into 2 such that the
+     * batch with the continuous records has only 1 acknowledge type in its array.
+     */
+    private List<ShareFetchRequestData.AcknowledgementBatch> maybeOptimiseAcknowledgementTypesForFetch(ShareFetchRequestData.AcknowledgementBatch currentShareFetchBatch) {
+        List<ShareFetchRequestData.AcknowledgementBatch> batches = new ArrayList<>();
+        if (currentShareFetchBatch == null) return batches;
+
+        long currentOffset = currentShareFetchBatch.firstOffset();
+        int currentStartIndex = 0;
+        int maxRecordsWithSameAcknowledgeType = 1;
+        for (int i = 1; i < currentShareFetchBatch.acknowledgeTypes().size(); i++) {
+            byte acknowledgeType = currentShareFetchBatch.acknowledgeTypes().get(i) == 0 ? ACKNOWLEDGE_TYPE_GAP : currentShareFetchBatch.acknowledgeTypes().get(i);
+
+            // If we have a continuous set of records with the same acknowledgement type exceeding the default count,
+            // then we optimise the batches to include only start and end offset and have only 1 acknowledge type in the array.
+            if (acknowledgeType == currentShareFetchBatch.acknowledgeTypes().get(i - 1) &&
```

Comment:
This test doesn't allow for the ACKNOWLEDGE_TYPE_GAP case.

Reply:
That case is also covered I think, because in the `getShareFetchBatches()` function, we check for null and add 0 if we see a gap(null) in the acknowledgements. So when it reaches the optimisation step, we already have a batch where the array of acknowledgement types will have 0 as the byte for gaps.